Feature/implement user module

Zlatonn opened the pull request:
**Features Included**
- implement user decorator for get user data from request
- implement user module
- separate password by role
- delete id at seeding

---
_Get list user service_

![messageImage_1756708770630](https://github.com/user-attachments/assets/b7203b9f-6a17-4f5c-973e-c8159680ec79)

_Get user by id_

![messageImage_1756708793727](https://github.com/user-attachments/assets/d2cb4809-7d89-47a9-8003-fb0e64b1ddc9)

_Create user_

![messageImage_1756708728166](https://github.com/user-attachments/assets/a1d46f80-5db8-4273-aba0-e5c94fe644aa)

_Update user_

![messageImage_1756708820940](https://github.com/user-attachments/assets/34d43c4c-86d7-4cbc-8a8a-c1efbaafccdf)

_Delete user_

![messageImage_1756708994565](https://github.com/user-attachments/assets/9549f491-83e8-4a36-a0c6-623a61995eae)

_Reset password_

![messageImage_1756708976170](https://github.com/user-attachments/assets/e62be109-53e6-4ae4-b1b9-6ea6b18bc026)

_Change password_

![messageImage_1756708894624](https://github.com/user-attachments/assets/7a3fb4a5-3e83-475f-a7dd-c5a6e1b0594c)

Zlatonn committed:
change name file skip-auth

- change skip-auth to skip-auth.decorator

Zlatonn committed:
implement user decorator

- for get user data from http request

Zlatonn committed:
implement user module

- create includes DTO
- create user module
- create user controllers
- create user services
- create pagination helper
- apply user module to app module

Zlatonn committed:
separate password by role

- update env
- update seeding
- update to resetPassword service

Zlatonn committed:
delete id for seeding data

Leekub1 commented on src/modules/user/dtos/update-user.dto.ts:
```diff
@@ -0,0 +1,24 @@
+import { IsEnum, IsNotEmpty, IsString } from 'class-validator';
+import { Role } from '@prisma/client';
+
+export class UpdateUserDto {
```
export class UpdateUserDTO extends PartialType(CreateUserDto)

แนะนำ PartialType ครับ

Zlatonn replied:
done kub ✅

Leekub1 commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    const userExists = await this.prismaService.user.findFirst({
+      where: { username, NOT: { id } },
```
findunique

Zlatonn replied:
done kub ✅

looksocii commented on src/modules/user/dtos/create-user.dto.ts:
```diff
@@ -0,0 +1,29 @@
+  lastName: string;
+
+  @IsNotEmpty()
+  @IsString()
```
length 10

Zlatonn replied:
อันนี้พี่ใช้ท่านี้ได้มะ พี่อยากได้ format นี้เฉยๆ 55555
<img width="501" height="146" alt="image" src="https://github.com/user-attachments/assets/86cf78ef-bb2a-48a7-8176-47ab87fdb957" />

looksocii replied:
ได้เหมือนกันครับ

looksocii commented on src/modules/user/dtos/update-user.dto.ts:
```diff
@@ -0,0 +1,24 @@
+  @IsString()
+  firstName: string;
+
+  @IsNotEmpty()
```
IsOptional ครับ ถ้า update ข้อมูลแค่บางฟิล

Zlatonn replied:
done kub ✅

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,265 @@
+        {
+          username: {
+            contains: search,
+            mode: 'insensitive',
```
type safety

Zlatonn replied:
น้องพี่ถามนิดนึง ตรงนี้พี่ลองพิมอย่างอื่นละมัน error อะน้องคือมันถือว่า safe แล้วมะ
<img width="337" height="203" alt="image" src="https://github.com/user-attachments/assets/b3b8a5c3-8c33-404c-9d34-ec662216c490" />

looksocii replied:
ok krub

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,265 @@
+          lastName: {
+            contains: search,
+            mode: 'insensitive',
+          },
```
ถ้าอยากพิมพ์ค้นหาแบบ full name จะค้นหาได้มั้ยคับ

Zlatonn replied:
ตอนนี้ไม่ได้น้องใน frontend  พี่กะจะแยก column firstname กะ lastname เอาครับ

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,265 @@
+    });
+
+    if (!user) {
+      throw new NotFoundException(`User is not found.`);
```
throw กลับไปบอกด้วยว่า id อะไรที่ไม่เจอ จะช่วยในการ debug ครับ

Zlatonn replied:
done kub ✅

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,265 @@
+    private readonly prismaService: PrismaService,
+    private readonly config: ConfigService,
+  ) {
+    this.saltRound = this.config.getOrThrow<string>('SALT_ROUND');
```
เป็น type number รึป่าวครับ
ตอนเอาไปใช้จะได้ไม่ต้องแปลงอีกครับ

Zlatonn replied:
done kub ✅

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,265 @@
+    const { username, firstName, lastName, phoneNumber, role } = req;
+
+    const userFound = await this.prismaService.user.findUnique({
+      where: { id },
```
select id
recheck all

Zlatonn replied:
done kub ✅

Zlatonn committed:
request change#1

- update type return of pagination helper functions
- update create-user DTO using matches phone format
- update update-user DTO using patialType, omitType
- update type of field saltRound to number
- add select for query at auth services & user services
- change error message at user services

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,271 @@
+        break;
+      default:
+        throw new BadRequestException('Unsupported role for password reset.');
+    }
```
ถ้าเพิ่ม role ใน enum อีก แล้ว switch case นี้จะเข้า default ตลอดรึป่าวครับ

Zlatonn replied:
done kub ✅

looksocii commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,271 @@
+    };
+
+    return await this.prismaService.user.update({
+      where: { id },
```
update จาก id ที่ find มาดีกว่าครับ

Zlatonn committed:
request change #2

- change condition for reset password user
- adjust format code
